src/models/image-map.ts

```diff
@@ -105,52 +105,54 @@ export const ImageMapModel = types
+        entry = self.images.get(url);
+        if (entry && snapshot.contentUrl) {
+          entry.contentUrl = snapshot.contentUrl;
         }
-        // create new entry
-        else {
-          entry = ImageMapEntry.create(snapshot);
-          self.images.set(url, entry);
-          // notify any listeners that the url is now available
-          if (self.listeners[url]) {
-            for (const id in self.listeners[url]) {
-              self.listeners[url][id]();
-            }
+        if (entry && snapshot.displayUrl) {
+          entry.displayUrl = snapshot.displayUrl;
+        }
+      }
+      // create new entry
+      else {
+        entry = ImageMapEntry.create(snapshot);
+        self.images.set(url, entry);
+        // notify any listeners that the url is now available
+        if (self.listeners[url]) {
+          for (const id in self.listeners[url]) {
+            self.listeners[url][id]();
           }
         }
-        self.syncContentUrl(url, entry!);
-
-        getImageDimensions(entry && entry.displayUrl || url)
-          .then(dimensions => {
-            const imageEntry = self.images.get(url);
-            if (imageEntry) {
-              self.setDimensions(url, dimensions.width, dimensions.height);
-              self.syncContentUrl(url, imageEntry);
-              resolve(imageEntry);
-            }
-          });
-      });
-    }
-  }))
-  .actions(self => ({
-    addPromise(url: string, promise: Promise<ImageMapEntrySnapshot>): Promise<ImageMapEntryType> {
-      return new Promise((resolve, reject) => {
-        promise.then(snapshot => {
-          resolve(self.addImage(url, snapshot));
-        });
-      });
+      }
+      self.syncContentUrl(url, entry!);
+
+      // If the getImageDimension image element never loads then we won't get
+      // past this line. 
+      // However the entry has already been added to the map, so there will be
+      // a dimensionless entry in this case.
+      // Also in most cases addImage is not called until the image has already been
+      // downloaded and displayUrl is actually a blob url.
+      // So it should be unlikely in these cases that getImageDimensions will fail.
+      const dimensions = await getImageDimensions(entry && entry.displayUrl || url);
+      const imageEntry = self.images.get(url);
+      if (!imageEntry) {
+        // This should really not happen, we just added an image entry above
+        // and there isn't a way to remove entries from the map
+        /* istanbul ignore next */
+        throw `missing image entry in cache for ${url}`;
+      }
+      self.setDimensions(url, dimensions.width, dimensions.height);
+      // If this addImage is called more than once there could be multiple
+      // of these promises running at once. And the last one to finish will 
+      // clobber the imageEntry.contentUrl entry that was there before.
+      // The last one to finish might not be the right one if there are multiple.
```

Comment:
Is there a TODO here, e.g. to track calls to `addImage()` and make sure that all calls for an image after the first return the original promise?

Reply:
Yeah I'm not sure what to do about this. In the next PR I was going update getImage to keep track of in progress requests. So maybe with that change this issue will get resolved.